### PR TITLE
CMDCT 261 - Admin Review Functionality

### PR DESCRIPTION
**Pull request by keeysnc**
### Description
https://jiraent.cms.gov/browse/CMDCT-261
This ticket completes the Admin Review functionality below:

1. Unlock functionality (Submitted reports only)

Enable Unlock and Approve Buttons
***  “Close” - Modal closes and user remains on the Review and Submit page
“Return to dashboard” takes the user to the [state specific admin dashboard](https://www.figma.com/file/oOJMUZerbafZefLQQ33GwI/MFP_Playground?type=design&node-id=3379%3A47897&mode=design&t=dF2v4U7WRZgbkAWH-1)

2. Approval Functionality

If the admin clicks the “Approve” button, a modal will appear that will require the admin type in “approve” (not case sensitive) and hit the “Approve” button inside the modal, which is validated on key stroke.
Update report to "Approved"
Redirect Admin to [state specific admin dashboard](https://www.figma.com/file/oOJMUZerbafZefLQQ33GwI/MFP_Playground?type=design&node-id=3379%3A47897&mode=design&t=dF2v4U7WRZgbkAWH-1)
Display updated status in [state specific admin dashboard](https://www.figma.com/file/oOJMUZerbafZefLQQ33GwI/MFP_Playground?type=design&node-id=3379%3A47897&mode=design&t=dF2v4U7WRZgbkAWH-1)

Unlock Report:
https://github.com/Enterprise-CMCS/macpro-mdct-mfp/assets/14514294/aa88d1a0-e71d-45a3-818a-bba5f80dd93f

Approve Report:

https://github.com/Enterprise-CMCS/macpro-mdct-mfp/assets/14514294/351fe210-8403-499f-8e8d-6a094ec9681f

### Related ticket(s)

CMDCT- Subtask- https://jiraent.cms.gov/browse/CMDCT-261
Main Ticket - https://jiraent.cms.gov/browse/CMDCT-2995

---
### How to test
To test you will need to mock a submitted report.
1. In create.ts 
Within the Item object add or update the properties *status: "Submitted"* and "locked: true" 
2. Sign back in as a state user and make sure the report generated has a Submitted status.
3. Sign out and sign in as a admin user
4. Select the state the report is in
4.  View report and go to the "Review and Submit" page ( you should see the admin review section below)
5. Test out the "Unlock" functionality then refresh dev environment and repeat steps 2 and 3 before testing the "Approve" button. ( Once you unlock the report the buttons disable and your redirected to the dashboard)

Here's the criteria for how the buttons and modals should function:
https://jiraent.cms.gov/browse/CMDCT-261

### Important updates

---
### Author checklist

- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

File: services/app-api/handlers/reports/approve.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  if (!getCurrentReport.body) {
+    return {
+      status: StatusCodes.NOT_FOUND,
+      body: error.NO_MATCHING_RECORD,
+    };
+  }
+
+  // if current report exists, parse for archived status
+  if (getCurrentReport?.body) {
```
Comment (gmrabian) [inline services/app-api/handlers/reports/approve.ts:31]: suggestion: do we need this second `if` wrapper? `getCurrentReport.body` must exist if the first `if` skipped, right?

**Review by gmrabian — COMMENTED** [review]
Functionality was all good for approve. Unlock behaved a little oddly when I did it from the dashboard. The error banner triggered but there was no failed network call, and the modal button to "return to dashboard" seems a little odd when we are already on that page, but I could also see interpreting it as a "stay here" kind of thing so that may not be a big deal

https://github.com/Enterprise-CMCS/macpro-mdct-mfp/assets/57802560/5ccc141f-efc4-4974-ba9e-5c4139a5f45a

File: services/app-api/handlers/reports/release.ts [hunk 2]
```diff
@@ -158,7 +158,7 @@ export const releaseReport = handler(async (event) => {
     fieldDataId: newFieldDataId,
     locked: false,
     previousRevisions,
-    status: "In progress",
+    status: "In Revision",
```
Comment (karla-vm) [inline services/app-api/handlers/reports/release.ts:161]: **nit:** just for consistency with the other statuses
```suggestion
    status: "In revision",
```

**Review by karla-vm — DISMISSED** [review]

File: services/app-api/handlers/reports/approve.ts (hunk 1 above)
Comment (keeysnc) [reply services/app-api/handlers/reports/approve.ts:31]: @gmrabian I can possibly remove it, not sure if there was a edge case that needed this. This code was pulled from the archive.ts

**Review by cassandradanger — COMMENTED** [review]
In the APPROVE modal, if the user overtypes "approve", the button should become disabled again 
Screenshot 2023-10-20 at 5 04 33 PM

**Comment by keeysnc** [comment]
> In the APPROVE modal, if the user overtypes "approve", the button should become disabled again Screenshot 2023-10-20 at 5 04 33 PM



> In the APPROVE modal, if the user overtypes "approve", the button should become disabled again Screenshot 2023-10-20 at 5 04 33 PM

Ah thanks for catching that

**Review by cassandradanger — COMMENTED** [review]
one tiny little design thing, since we have more time with the new deadline.
There should be more space between the copy and the button in the unlock modal
Screenshot 2023-10-21 at 9 26 34 AM

**Review by cassandradanger — APPROVED** [review]

**Review by karla-vm — APPROVED** [review]
💯

**Pull request merged**